### PR TITLE
⭐️ GCP Resource Pack

### PR DESCRIPTION
**Pull request by chris-rock**
The the following resources will be included with this resource pack:

- [x] compute.instance
- [x] compute.disk
- [x] compute.snapshot
- [x] compute.image
- [x] compute.network
- [x] compute.router
- [x] compute.zone
- [x] compute.region
- [x] dns.managedzone
- [x] dns.record
- [x] dns.policy
- [x] bigquery.instance
- [x] bigquery.table
- [x] bigquery.dataset
- [x] recommendations

## Check if individual services are enabled

```
cnquery> gcp.service(name: "compute.googleapis.com").enabled
[ok] value: true
```

## List of GCP Instances

```
cnquery> gcp.compute.instances
gcp.compute.instances: [
  0: gcp.compute.instance name="centos7"
  1: gcp.compute.instance name="centos7-2020-09-1"
  2: gcp.compute.instance name="centos7-latest"
  3: gcp.compute.instance name="rhel6"
  4: gcp.compute.instance name="rhel7-2"
  5: gcp.compute.instance name="rhel8"
  6: gcp.compute.instance name="suse"
  7: gcp.compute.instance name="ubuntu2004"
  8: gcp.compute.instance name="windows"
]
cnquery> 
```

## List all instances with name and attached disks

```
gcp.compute.instances { name disks { * } }
  0: {
    disks: [
      0: {
        mode: "READ_WRITE"
        forceAttach: false
        deviceName: "centos7"
        type: "PERSISTENT"
        licenses: [
          0: "https://www.googleapis.com/compute/v1/projects/centos-cloud/global/licenses/centos-7"
        ]
        index: 0
        guestOsFeatures: [
          0: "UEFI_COMPATIBLE"
        ]
        boot: true
        diskSizeGb: 20
        projectId: "example-123"
        interface: "SCSI"
        architecture: ""
        autoDelete: true
        id: "1234567891234567890/0"
      }
    ]
    name: "centos7"
  }
```

## List VM snapshots

```
cnquery> gcp.compute.snapshots
gcp.compute.snapshots: [
  0: gcp.compute.snapshot name="suse12"
  1: gcp.compute.snapshot name="suse15"
]
```

## List all VM images

```
cnquery> gcp.compute.images
gcp.compute.images: [
  0: gcp.compute.image id="1234567891234567890" name="gcp-ubuntu-2004-secure-base-20220704145254"
  1: gcp.compute.image id="1234567891234567891" name="gcp-ubuntu-2004-secure-base-20220704145847"
]
```

## List all networks

```
cnquery> gcp.compute.networks
gcp.compute.networks: [
  0: gcp.compute.network name="default"
]
```

## List all firewalls

```
cnquery> gcp.compute.firewalls
gcp.compute.firewalls: [
  0: gcp.compute.firewall name="default-allow-http"
  1: gcp.compute.firewall name="default-allow-https"
  2: gcp.compute.firewall name="default-allow-icmp"
  3: gcp.compute.firewall name="default-allow-internal"
  4: gcp.compute.firewall name="default-allow-rdp"
  5: gcp.compute.firewall name="default-allow-ssh"
]
```

## List all zones

```
cnquery> gcp.compute.zones
gcp.compute.zones: [
  0: gcp.compute.zone name="us-east1-b"
  1: gcp.compute.zone name="us-east1-c"
  2: gcp.compute.zone name="us-east1-d"
  3: gcp.compute.zone name="us-east4-c"
  4: gcp.compute.zone name="us-east4-b"
  5: gcp.compute.zone name="us-east4-a"
  6: gcp.compute.zone name="us-central1-c"
  7: gcp.compute.zone name="us-central1-a"
  8: gcp.compute.zone name="us-central1-f"
  9: gcp.compute.zone name="us-central1-b"
  10: gcp.compute.zone name="us-west1-b"
  11: gcp.compute.zone name="us-west1-c"
  12: gcp.compute.zone name="us-west1-a"
  13: gcp.compute.zone name="europe-west4-a"
  14: gcp.compute.zone name="europe-west4-b"
  15: gcp.compute.zone name="europe-west4-c"
  16: gcp.compute.zone name="europe-west1-b"
  17: gcp.compute.zone name="europe-west1-d"
  18: gcp.compute.zone name="europe-west1-c
  ```
  
  
  ## List DNS managed zones
  
  ```
  cnquery> gcp.dns.managedZones
  ```
  
  ## List DNS managed zones with record sets
  
  ```
  cnquery> gcp.dns.managedZones { name recordSets { * } }
 ```
 
 
 ## List Recommendation for projects

```
cnquery> gcp.project.recommendations { name primaryImpact  } 
gcp.project.recommendations: [
  0: {
    name: "Save cost by stopping Idle VM 'example-dev'."
    primaryImpact: {
      Projection: {
        CostProjection: {
          cost: {
            currency_code: "USD"
            nanos: -25987334.000000
            units: -24.000000
          }
          duration: {
            seconds: 2592000.000000
          }
        }
      }
      category: 1.000000
    }
  }
]
```

### CODE REVIEW

File: resources/packs/gcp/compute.go [hunk 1]
```diff
@@ -34,77 +87,1138 @@ func (g *mqlGcpCompute) GetInstances() ([]interface{}, error) {
...
+func (g *mqlGcpComputeZone) id() (string, error) {
+	id, err := g.Id()
+	if err != nil {
+		return "", err
+	}
+	return "gcp.compute.zone/" + id, nil
+}
+
+func (g *mqlGcpComputeZone) GetRegion() (interface{}, error) {
+	// TODO: implement
```
Comment (chris-rock) [inline resources/packs/gcp/compute.go:138]: the reference handling will be implemented in a followup PR

File: motor/providers/google/clients.go [hunk 2]
```diff
@@ -11,6 +11,22 @@ import (
 	"google.golang.org/api/transport"
 )
 
+func (t *Provider) Credentials(scopes ...string) (*googleoauth.Credentials, error) {
+	ctx := context.Background()
+
+	if t.serviceAccountSubject == "" {
+		return googleoauth.FindDefaultCredentials(ctx, scopes...)
```
Comment (preslavgerchev) [inline motor/providers/google/clients.go]: i assume this uses the gcloud cli somehow? can we add a debug log here (or where this is called), indicating that we're falling back to cli auth? would make it easier for debugging purposes if we need that

Comment (chris-rock) [reply motor/providers/google/clients.go]: Great call, lets do that

File: resources/packs/gcp/recommendations.go [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+	}
+
+	ctx := context.Background()
+	c, err := recommender.NewClient(ctx, option.WithCredentials(credentials))
+	if err != nil {
+		log.Info().Err(err).Msg("could not create client")
+		return nil, err
+	}
+
+	parent := fmt.Sprintf("projects/%s/locations/%s/recommenders/%s", projectId, "us-central1-a", "google.compute.instance.IdleResourceRecommender")
```
Comment (preslavgerchev) [inline resources/packs/gcp/recommendations.go]: the location here is harcoded, can it be parameterized?

**Comment by scottford-io** [comment]
@chris-rock Have you seen this? 

```javascript
gcp.compute.disks {*}
x error="failed to create resource 'gcp.compute.disk': Failed to initialize \"gcp.compute.disk\", its \"guestOsFeatures\" argument has the wrong type (expected type \"[]interface{}\")"
gcp.compute.disks: []
```

**Comment by scottford-io** [comment]
```
cnquery> gcp.compute.snapshots {*}
Query encountered errors:
1 error occurred:
	* not implemented
```


```javascript
cnquery> gcp.compute.snapshots {*}
Query encountered errors:
1 error occurred:
	* not implemented
gcp.compute.snapshots: [
  0: {
    diskSizeGb: 10
    storageBytesStatus: "UP_TO_DATE"
    sourceDisk: not implemented
    description: ""
    storageBytes: 554335808
    snapshotType: ""
    status: "READY"
    name: "suse12"
    licenses: [
      0: "https://www.googleapis.com/compute/v1/projects/suse-cloud/global/licenses/sles-12"
    ]
```

**Comment by scottford-io** [comment]
### Subnetworks not implemented

```
cnquery> gcp.compute.networks {*}
Query encountered errors:
1 error occurred:
	* not implemented
gcp.compute.networks: [
  0: {
    description: "Default network for the project"
    created: 2019-12-17 05:03:16.947 -0800 PST
    subnetworks: not implemented
    name: "default"
    networkFirewallPolicyEnforcementOrder: "AFTER_CLASSIC_FIREWALL"
    mtu: 0
    gatewayIPv4: ""
    autoCreateSubnetworks: true
    routingMode: "REGIONAL"
    peerings: []
    enableUlaInternalIpv6: false
    id: "4155011992112164091"
  }
]
```

**Comment by scottford-io** [comment]
### Region not implemented

```
cnquery> gcp.compute.zones {*}
Query encountered errors:
1 error occurred:
	* not implemented
gcp.compute.zones: [
  0: {
    id: "2231"
    description: "us-east1-b"
    created: 719528 days
    region: not implemented
    name: "us-east1-b"
    status: "UP"
  }
  ```

**Comment by scottford-io** [comment]
```
cnquery> gcp.dns.managedZones
Query encountered errors:
googleapi: Error 403: Cloud DNS API has not been used in project 458067389847 before or it is disabled. Enable it by visiting https://console.developers.google.com/apis/api/dns.googleapis.com/overview?project=458067389847 then retry. If you enabled this API recently, wait a few minutes for the action to propagate to our systems and retry.
Details:
[
  {
    "@type": "type.googleapis.com/google.rpc.Help",
    "links": [
      {
        "description": "Google developers console API activation",
        "url": "https://console.developers.google.com/apis/api/dns.googleapis.com/overview?project=458067389847"
      }
    ]
  },
  {
    "@type": "type.googleapis.com/google.rpc.ErrorInfo",
    "domain": "googleapis.com",
    "metadata": {
      "consumer": "projects/458067389847",
      "service": "dns.googleapis.com"
    },
    "reason": "SERVICE_DISABLED"
  }
]
, accessNotConfigured
gcp.dns.managedZones: no data available
```

**Comment by chris-rock** [comment]
@scottford-io As mentioned, those references are follow-up work. I can take them out of the resource definitions for now

**Comment by scottford-io** [comment]
### help for gcp.bigquery does not work...

```
cnquery> help gcp.bigquery
```

**Comment by chris-rock** [comment]
> ### help for gcp.bigquery does not work...
> 
> ```
> cnquery> help gcp.bigquery
> ```

This issue is not related to this improvement, I create a followup issue for that finding

**Comment by scottford-io** [comment]
## cannot find resource factory for 'gcp.resourcemanager'

```
cnquery> gcp.resourcemanager
Query encountered errors:
cannot find resource factory for 'gcp.resourcemanager'
gcp.resourcemanager: no data available
```

**Comment by scottford-io** [comment]
### failed to create resource 'gcp.service': "gcp.service" failed: no value provided for static field "name"

```
cnquery> gcp.service
Query encountered errors:
failed to create resource 'gcp.service': "gcp.service" failed: no value provided for static field "name"
gcp.service: no data available
```

**Comment by scottford-io** [comment]
@chris-rock amazing to see all of these new resources!!!! ⭐  

I connected a shell to one of the GCP projects I have access to and tested things out. I just raised up any issues I found. I am guessing that a number of them you are already addressing, or they are a non-issue. Hopefully this helps!

**Comment by chris-rock** [comment]
> ### failed to create resource 'gcp.service': "gcp.service" failed: no value provided for static field "name"
> 
> ```
> cnquery> gcp.service
> Query encountered errors:
> failed to create resource 'gcp.service': "gcp.service" failed: no value provided for static field "name"
> gcp.service: no data available
> ```

You cannot call `gcp.service` directly. It will always need a service name as input. Eg:

```
gcp.service(name: "compute.googleapis.com").enabled
``` 

will work.


- `gcp.resourcemanager` is a group and cannot be called directly
- I am going to remove all the placeholder for relationships for followup improvements

**Comment by chris-rock** [comment]
> gcp.dns.managedZones

I am not sure what we can do better here. This just needs to be activated.

**Comment by chris-rock** [comment]
> ```js
> gcp.compute.disks {*}
> ```

Fixed

File: motor/providers/google/clients.go (hunk 2 above)
Comment (chris-rock) [reply motor/providers/google/clients.go]: I added a debug message for this case

File: resources/packs/gcp/recommendations.go (hunk 3 above)
Comment (chris-rock) [reply resources/packs/gcp/recommendations.go]: I updated the recommender handling to iterate over all zones and recommenders.

**Review by preslavgerchev — APPROVED** [review]

**Pull request merged**